README
.. -*- restructuredtext -*-

##########################################
tweetstream - Simple twitter streaming API
##########################################

Introduction
------------

tweetstream provides two classes, SampleStream and FollowStream, that can be
used to get tweets from Twitter's streaming API. An instance of one of the
classes can be used as an iterator. In addition to fetching tweets, the 
object keeps track of the number of tweets collected and the rate at which
tweets are received.

SampleStream delivers a sample of all tweets. FilterStream delivers
tweets that match one or more criteria. Note that it's not possible
to get all tweets without access to the "firehose" stream, which
is not currently avaliable to the public.

Twitter's documentation about the streaming API can be found here:
http://dev.twitter.com/pages/streaming_api_methods .

**Note** that the API is blocking. If for some reason data is not immediatly
available, calls will block until enough data is available to yield a tweet.

Examples
--------

Printing incoming tweets:

>>> stream = tweetstream.SampleStream("username", "password")
>>> for tweet in stream:
...     print tweet


The stream object can also be used as a context, as in this example that
prints the author for each tweet as well as the tweet count and rate:

>>> with tweetstream.SampleStream("username", "password") as stream
...     for tweet in stream:
...         print "Got tweet from %-16s\t( tweet %d, rate %.1f tweets/sec)" % (
...                 tweet["user"]["screen_name"], stream.count, stream.rate )


Stream objects can raise ConnectionError or AuthenticationError exceptions:

>>> try:
...     with tweetstream.TweetStream("username", "password") as stream
...         for tweet in stream:
...             print "Got tweet from %-16s\t( tweet %d, rate %.1f tweets/sec)" % (
...                     tweet["user"]["screen_name"], stream.count, stream.rate )
... except tweetstream.ConnectionError, e:
...     print "Disconnected from twitter. Reason:", e.reason

To get tweets that match specific criteria, use the FilterStream. FilterStreams
take three keyword arguments: "locations", "follow" and "track".

Locations are a list of bounding boxes in which geotagged tweets should originate.
The argument should be an iterable of longitude/latitude pairs.

Track specifies keywords to track. The argument should be an iterable of
strings.

Follow returns statuses that reference given users. Argument should be an iterable
of twitter user IDs. The IDs are userid ints, not the screen names. 

>>> words = ["opera", "firefox", "safari"]
>>> people = [123,124,125]
>>> locations = ["-122.75,36.8", "-121.75,37.8"]
>>> with tweetstream.FilterStream("username", "password", track=words,
...                               follow=people, locations=locations) as stream
...     for tweet in stream:
...         print "Got interesting tweet:", tweet


Deprecated classes
------------------

tweetstream used to contain the classes TweetStream, FollowStream, TrackStream
LocationStream and ReconnectingTweetStream. These were deprecated when twitter
changed its API end points. The same functionality is now available in
SampleStream and FilterStream. The deprecated methods will emit a warning when
used, but will remain functional for a while longer.


Changelog
---------

See the CHANGELOG file

Contact
-------

The author is Rune Halvorsen <runefh@gmail.com>. The project resides at
http://bitbucket.org/runeh/tweetstream . If you find bugs, or have feature
requests, please report them in the project site issue tracker. Patches are
also very welcome.

Contributors
------------

- Rune Halvorsen
- Christopher Schierkolk

License
-------

This software is licensed under the ``New BSD License``. See the ``LICENCE``
file in the top distribution directory for the full license text.


CHANGELOG
0.1

 - Initial version

0.2

 - Improved error handling
 - Added AuthenticationError and ConnectionError exceptions
 - Added ReconnectingTweetStream class that supports automatically
   reconnecting if the connection is dropped

0.3

 - Fixed bugs in authtentication
 - Added TrackStream and FollowStream classes
 - Added list of endpoint names, and made them legal values for the url arg

0.3.1

 - Added lots of tests
 - Added proper handling of keepalive newlines
 - Improved handling of closing streams
 - Added missing anyjson dependency to setup
 - Fixed bug where newlines and malformed content were counted as a tweet

0.3.2

 - This release was skipped over, due to maintainer brainfart.

0.3.3

 - Fixed setup.py so it wont attempt to load modules that aren't installed
   yet. Fixes installation issue.

0.3.4

 - Updated to latest twitter streaming urls
 - Fixed a bug where we tried to call a method on None

0.3.5

 - Removed a spurious print statement left over from debugging
 - Introduced common base class for all tweetstream exceptions
 - Make sure we raise a sensible error on 404. Include url in desc of that error

0.3.6

 - Added LocationStream class for filtering on location bounding boxes.

1.0.0

 - Changed API to match latest twitter endpoints. This adds SampleStream and
   FilterStream and deprecates TweetStream, FollowStream, LocationStream,
   TrackStream and ReconnectingTweetStream.

1.1.0

 - Fixed issues #2 and #12, related to low volume streams not yielding tweets
   until a relatively large buffer was filled. This meant that tweets would
   arrive in bunches, not immediatly.
 - Switched to HTTPS urls for streams. Twitter will switch off HTTP streams
   on 29. sept. 2011.
 - Added support for Python 3

1.1.1

 - Fixed issue #16. Odd case where python_version_tuple was returning ints
   rather than the strings the docs promis. Make sure we always cast to int.

1.2

 - Added "raw" kwarg to stream classes. Returns raw tweets rather than
   decoding the JSON.

 - New, fine grained, disconnect exceptions. Adds a ReconnectableError
   of which there are three subclases for different reconnect behaviours.

 

tests/test_tweetstream.py
import contextlib
import threading
import time

from tweetstream import TweetStream, FollowStream, TrackStream, LocationStream
from tweetstream import ConnectionError, AuthenticationError, SampleStream, FilterStream

import pytest
from pytest import raises
slow = pytest.mark.slow

from servercontext import test_server

single_tweet = r"""{"in_reply_to_status_id":null,"in_reply_to_user_id":null,"favorited":false,"created_at":"Tue Jun 16 10:40:14 +0000 2009","in_reply_to_screen_name":null,"text":"record industry just keeps on amazing me: http:\/\/is.gd\/13lFo - $150k per song you've SHARED, not that somebody has actually DOWNLOADED.","user":{"notifications":null,"profile_background_tile":false,"followers_count":206,"time_zone":"Copenhagen","utc_offset":3600,"friends_count":191,"profile_background_color":"ffffff","profile_image_url":"http:\/\/s3.amazonaws.com\/twitter_production\/profile_images\/250715794\/profile_normal.png","description":"Digital product developer, currently at Opera Software. My tweets are my opinions, not those of my employer.","verified_profile":false,"protected":false,"favourites_count":0,"profile_text_color":"3C3940","screen_name":"eiriksnilsen","name":"Eirik Stridsklev N.","following":null,"created_at":"Tue May 06 12:24:12 +0000 2008","profile_background_image_url":"http:\/\/s3.amazonaws.com\/twitter_production\/profile_background_images\/10531192\/160x600opera15.gif","profile_link_color":"0099B9","profile_sidebar_fill_color":"95E8EC","url":"http:\/\/www.stridsklev-nilsen.no\/eirik","id":14672543,"statuses_count":506,"profile_sidebar_border_color":"5ED4DC","location":"Oslo, Norway"},"id":2190767504,"truncated":false,"source":"<a href=\"http:\/\/widgets.opera.com\/widget\/7206\">Twitter Opera widget<\/a>"}""" + "\r"


def parameterized(funcarglist):
    def wrapper(function):
        function.funcarglist = funcarglist
        return function
    return wrapper

def pytest_generate_tests(metafunc):
    for funcargs in getattr(metafunc.function, 'funcarglist', ()):
        metafunc.addcall(funcargs=funcargs)


streamtypes = [
    dict(cls=TweetStream, args=[], kwargs=dict()),
    dict(cls=SampleStream, args=[], kwargs=dict()),
    dict(cls=FilterStream, args=[], kwargs=dict(track=("test",))),
    dict(cls=FollowStream, args=[[1, 2, 3]], kwargs=dict()),
    dict(cls=TrackStream, args=["opera"], kwargs=dict()),
    dict(cls=LocationStream, args=["123,4321"], kwargs=dict())
]


@parameterized(streamtypes)
def test_bad_auth(cls, args, kwargs):
    """Test that the proper exception is raised when the user could not be
    authenticated"""
    def auth_denied(request):
        request.send_error(401)

    with raises(AuthenticationError):
        with test_server(handler=auth_denied, methods=("post", "get"), port="random") as server:
            stream = cls("user", "passwd", *args, url=server.baseurl)
            for e in stream: pass


@parameterized(streamtypes)
def test_404_url(cls, args, kwargs):
    """Test that the proper exception is raised when the stream URL can't be
    found"""
    def not_found(request):
        request.send_error(404)

    with raises(ConnectionError):
        with test_server(handler=not_found, methods=("post", "get"), port="random") as server:
            stream = cls("user", "passwd", *args, url=server.baseurl)
            for e in stream: pass


@parameterized(streamtypes)
def test_bad_content(cls, args, kwargs):
    """Test error handling if we are given invalid data"""
    def bad_content(request):
        for n in xrange(10):
            # what json we pass doesn't matter. It's not verifying the
            # strcuture, only checking that it's parsable
            yield "[1,2,3]\r"
        yield "[1,2, I need no stinking close brace\r"
        yield "[1,2,3]\r"


    with raises(ConnectionError):
        with test_server(handler=bad_content, methods=("post", "get"), port="random") as server:
            stream = cls("user", "passwd", *args, url=server.baseurl)
            for tweet in stream:
                pass


@parameterized(streamtypes)
def test_closed_connection(cls, args, kwargs):
    """Test error handling if server unexpectedly closes connection"""
    cnt = 1000
    def bad_content(request):
        for n in xrange(cnt):
            # what json we pass doesn't matter. It's not verifying the
            # strcuture, only checking that it's parsable
            yield "[1,2,3]\r"

    with raises(ConnectionError):
        with test_server(handler=bad_content, methods=("post", "get"), port="random") as server:
            stream = cls("foo", "bar", *args, url=server.baseurl)
            for tweet in stream:
                pass


@parameterized(streamtypes)
def test_bad_host(cls, args, kwargs):
    """Test behaviour if we can't connect to the host"""
    with raises(ConnectionError):
        stream = cls("username", "passwd", *args, url="http://wedfwecfghhreewerewads.foo")
        stream.next()


@parameterized(streamtypes)
def smoke_test_receive_tweets(cls, args, kwargs):
    """Receive 100k tweets and disconnect (slow)"""
    total = 100000

    def tweetsource(request):
        while True:
            yield single_tweet + "\n"

    with test_server(handler=tweetsource, methods=("post", "get"), port="random") as server:
        stream = cls("foo", "bar", *args, url=server.baseurl)
        for tweet in stream:
            if stream.count == total:
                break


@parameterized(streamtypes)
def test_keepalive(cls, args, kwargs):
    """Make sure we behave sanely when there are keepalive newlines in the
    data recevived from twitter"""
    def tweetsource(request):
        yield single_tweet+"\n"
        yield "\n"
        yield "\n"
        yield single_tweet+"\n"
        yield "\n"
        yield "\n"
        yield "\n"
        yield "\n"
        yield "\n"
        yield "\n"
        yield "\n"
        yield single_tweet+"\n"
        yield "\n"


    with test_server(handler=tweetsource, methods=("post", "get"), port="random") as server:
        stream = cls("foo", "bar", *args, url=server.baseurl)
        try:
            for tweet in stream:
                pass
        except ConnectionError:
            assert stream.count == 3, "Got %s, wanted 3" % stream.count
        else:
            assert False, "Didn't handle keepalive"


@slow
@parameterized(streamtypes)
def test_buffering(cls, args, kwargs):
    """Test if buffering stops data from being returned immediately.
    If there is some buffering in play that might mean data is only returned
    from the generator when the buffer is full. If buffer is bigger than a
    tweet, this will happen. Default buffer size in the part of socket lib
    that enables readline is 8k. Max tweet length is around 3k."""

    def tweetsource(request):
        yield single_tweet+"\n"
        time.sleep(2)
        # need to yield a bunch here so we're sure we'll return from the
        # blocking call in case the buffering bug is present.
        for n in xrange(100):
            yield single_tweet+"\n"

    with test_server(handler=tweetsource, methods=("post", "get"), port="random") as server:
        stream = cls("foo", "bar", *args, url=server.baseurl)
        start = time.time()
        stream.next()
        first = time.time()
        diff = first - start
        assert diff < 1, "Getting first tweet took too long! %i > 1" % (diff)


servercontext.py
import threading
import contextlib
import time
import os
import socket
import random
from functools import partial
from inspect import isclass
from BaseHTTPServer import HTTPServer, BaseHTTPRequestHandler
from SimpleHTTPServer import SimpleHTTPRequestHandler
from SocketServer import BaseRequestHandler


class ServerError(Exception):
    pass


class ServerContext(object):
    """Context object with information about a running test server."""

    def __init__(self, address, port):
        self.address = address or "localhost"
        self.port = port

    @property
    def baseurl(self):
        return "http://%s:%s" % (self.address, self.port)

    def __str__(self):
        return "<ServerContext %s >" % self.baseurl

    __repr__ = __str__


class _SilentSimpleHTTPRequestHandler(SimpleHTTPRequestHandler):

    def __init__(self, *args, **kwargs):
        self.logging = kwargs.get("logging", False)
        SimpleHTTPRequestHandler.__init__(self, *args, **kwargs)

    def log_message(self, *args, **kwargs):
        if self.logging:
            SimpleHTTPRequestHandler.log_message(self, *args, **kwargs)


class _TestHandler(BaseHTTPRequestHandler):
    """RequestHandler class that handles requests that use a custom handler
    callable."""

    def __init__(self, handler, methods, *args, **kwargs):
        self._handler = handler
        self._methods = methods
        self._response_sent = False
        self._headers_sent = False
        self.logging = kwargs.get("logging", False)
        BaseHTTPRequestHandler.__init__(self, *args, **kwargs)

    def log_message(self, *args, **kwargs):
        if self.logging:
            BaseHTTPRequestHandler.log_message(self, *args, **kwargs)


    def send_response(self, *args, **kwargs):
        self._response_sent = True
        BaseHTTPRequestHandler.send_response(self, *args, **kwargs)

    def end_headers(self, *args, **kwargs):
        self._headers_sent = True
        BaseHTTPRequestHandler.end_headers(self, *args, **kwargs)

    def _do_whatever(self):
        """Called in place of do_METHOD"""
        data = self._handler(self)

        if hasattr(data, "next"):
            # assume it's something supporting generator protocol
            self._handle_with_iterator(data)
        else:
            # Nothing more to do then.
            pass


    def __getattr__(self, name):
        if name.startswith("do_") and name[3:].lower() in self._methods:
            return self._do_whatever
        else:
            # fixme instance or class?
            raise AttributeError(name)

    def _handle_with_iterator(self, iterator):
        for data in iterator:
            if not self.server.server_thread.running:
                return

            if not self._response_sent:
                self.send_response(200)
            if not self._headers_sent:
                self.end_headers()

            self.wfile.write(data)
            # flush immediatly. We may want to do trickling writes
            # or something else tha trequires bypassing normal caching
            self.wfile.flush()

class _TestServerThread(threading.Thread):
    """Thread class for a running test server"""

    def __init__(self, handler, methods, cwd, port, address):
        threading.Thread.__init__(self)
        self.startup_finished = threading.Event()
        self._methods = methods
        self._cwd = cwd
        self._orig_cwd = None
        self._handler = self._wrap_handler(handler, methods)
        self._setup()
        self.running = True
        self.serverloc = (address, port)
        self.error = None

    def _wrap_handler(self, handler, methods):
        if isclass(handler) and issubclass(handler, BaseRequestHandler):
            return handler # It's OK. user passed in a proper handler
        elif callable(handler):
            return partial(_TestHandler, handler, methods)
            # it's a callable, so wrap in a req handler
        else:
            raise ServerError("handler must be callable or RequestHandler")

    def _setup(self):
        if self._cwd != "./":
            self._orig_cwd = os.getcwd()
            os.chdir(self._cwd)

    def _init_server(self):
        """Hooks up the server socket"""
        try:
            if self.serverloc[1] == "random":
                retries = 10 # try getting an available port max this many times
                while True:
                    try:
                        self.serverloc = (self.serverloc[0],
                                          random.randint(1025, 49151))
                        self._server = HTTPServer(self.serverloc, self._handler)
                    except socket.error:
                        retries -= 1
                        if not retries: # not able to get a port.
                            raise
                    else:
                        break
            else: # use specific port. this might throw, that's expected
                self._server = HTTPServer(self.serverloc, self._handler)
        except socket.error, e:
            self.running = False
            self.error = e
            # set this here, since we'll never enter the serve loop where
            # it is usually set:
            self.startup_finished.set()
            return

        self._server.allow_reuse_address = True # lots of tests, same port
        self._server.timeout = 0.1
        self._server.server_thread = self


    def run(self):
        self._init_server()

        while self.running:
            self._server.handle_request() # blocks for self.timeout secs
            # First time this falls through, signal the parent thread that
            # the server is ready for incomming connections
            if not self.startup_finished.is_set():
                self.startup_finished.set()

        self._cleanup()

    def stop(self):
        """Stop the server and attempt to make the thread terminate.
        This happens async but the calling code can check periodically
        the isRunning flag on the thread object.
        """
        # actual stopping happens in the run method
        self.running = False

    def _cleanup(self):
        """Do some rudimentary cleanup."""
        if self._orig_cwd:
            os.chdir(self._orig_cwd)


@contextlib.contextmanager
def test_server(handler=_SilentSimpleHTTPRequestHandler, port=8514,
                address="", methods=("get", "head"), cwd="./"):
    """Context that makes available a web server in a separate thread"""
    thread = _TestServerThread(handler=handler, methods=methods, cwd=cwd,
                               port=port, address=address)
    thread.start()

    # fixme: should this be daemonized? If it isn't it will block the entire
    # app, but that should never happen anyway..
    thread.startup_finished.wait()

    if thread.error: # startup failed! Bail, throw whatever the server did
        raise thread.error

    exc = None
    try:
        yield ServerContext(*thread.serverloc)
    except Exception, exc:
        pass
    thread.stop()
    thread.join(5) # giving it a lot of leeway. should never happen

    if exc:
        raise exc

    # fixme: this takes second priorty after the internal exception but would
    # still be nice to signal back to calling code.

    if thread.isAlive():
        raise Warning("Test server could not be stopped")


tweetstream/__init__.py
"""Simple access to Twitter's streaming API"""

VERSION = (1, 1, 1)
__version__ = ".".join(map(str, VERSION[0:3])) + "".join(VERSION[3:])
__author__ = "Rune Halvorsen"
__contact__ = "runefh@gmail.com"
__homepage__ = "http://bitbucket.org/runeh/tweetstream/"
__docformat__ = "restructuredtext"

# -eof meta-


"""
 .. data:: USER_AGENT

     The default user agent string for stream objects
"""

USER_AGENT = "TweetStream %s" % __version__


class TweetStreamError(Exception):
    """Base class for all tweetstream errors"""
    pass


class ConnectionError(TweetStreamError):
    """Raised when there are network problems. This means when there are
    dns errors, network errors, twitter issues"""

    def __init__(self, reason, details=None):
        self.reason = reason
        self.details = details

    def __str__(self):
        return '<%s %s>' % (self.__class__.__name__, self.reason)


# The following exceptions track the various types of reconnection guidance
# given in the Twitter API documentation
# (https://dev.twitter.com/docs/streaming-api/concepts#connecting). In the
# code, if it's unclear which kind of reconnection is appropriate, the more
# conservative exception is raised.

class ReconnectError(ConnectionError):
    """Base class of the reconnectable connection errors."""
    pass

class ReconnectImmediatelyError(ReconnectError):
    '''From Twitter docs: "Once a valid connection connection drops, reconnect
    immediately."'''
    pass

class ReconnectLinearlyError(ReconnectError):
    '''From Twitter docs: "When a network error (TCP/IP level) is encountered,
    back off linearly. Perhaps start at 250 milliseconds and cap at 16
    seconds. Network layer problems are generally transitory and tend to clear
    quickly."'''
    pass

class ReconnectExponentiallyError(ReconnectError):
    '''From Twitter docs: "When a HTTP error (> 200) is returned, back off
    exponentially. Perhaps start with a 10 second wait, double on each
    subsequent failure, and finally cap the wait at 240 seconds. Consider
    sending an alert to a human operator after multiple HTTP errors, as there
    is probably a client configuration issue that is unlikely to be resolved
    without human intervention."'''
    pass


# Parent class per Twitter error handling guidance noted above.
class AuthenticationError(ReconnectExponentiallyError):
    """Exception raised if the username/password is not accepted"""
    pass


from .streamclasses import SampleStream, FilterStream
from .deprecated import FollowStream, TrackStream, LocationStream, TweetStream, ReconnectingTweetStream


tweetstream/deprecated.py
from .streamclasses import FilterStream, SampleStream

class DeprecatedStream(FilterStream):
    def __init__(self, *args, **kwargs):
        import warnings
        warnings.warn("%s is deprecated. Use FilterStream instead" % self.__class__.__name__, DeprecationWarning)
        super(DeprecatedStream, self).__init__(*args, **kwargs)


class FollowStream(DeprecatedStream):
    def __init__(self, username, password, follow, catchup=None, url=None, slow=False):
        super(FollowStream, self).__init__(username, password, follow=follow, catchup=catchup, url=url, slow=slow)


class TrackStream(DeprecatedStream):
    def __init__(self, username, password, track, catchup=None, url=None, slow=False):
        super(TrackStream, self).__init__(username, password, track=track, catchup=catchup, url=url, slow=slow)


class LocationStream(DeprecatedStream):
    def __init__(self, username, password, locations, catchup=None, url=None, slow=False):
        super(LocationStream, self).__init__(username, password, locations=locations, catchup=catchup, url=url, slow=slow)


class TweetStream(SampleStream):
    def __init__(self, *args, **kwargs):
        import warnings
        warnings.warn("%s is deprecated. Use SampleStream instead" % self.__class__.__name__, DeprecationWarning)
        SampleStream.__init__(self, *args, **kwargs)


class ReconnectingTweetStream(TweetStream):
    """TweetStream class that automatically tries to reconnect if the
    connecting goes down. Reconnecting, and waiting for reconnecting, is
    blocking.

    :param username: See :TweetStream:

    :param password: See :TweetStream:

    :keyword url: See :TweetStream:

    :keyword reconnects: Number of reconnects before a ConnectionError is
        raised. Default is 3

    :error_cb: Optional callable that will be called just before trying to
        reconnect. The callback will be called with a single argument, the
        exception that caused the reconnect attempt. Default is None

    :retry_wait: Time to wait before reconnecting in seconds. Default is 5

    """

    def __init__(self, username, password, url="sample",
                 reconnects=3, error_cb=None, retry_wait=5):
        self.max_reconnects = reconnects
        self.retry_wait = retry_wait
        self._reconnects = 0
        self._error_cb = error_cb
        TweetStream.__init__(self, username, password, url=url)

    def next(self):
        while True:
            try:
                return TweetStream.next(self)
            except ConnectionError, e:
                self._reconnects += 1
                if self._reconnects > self.max_reconnects:
                    raise ConnectionError("Too many retries")

                # Note: error_cb is not called on the last error since we
                # raise a ConnectionError instead
                if  callable(self._error_cb):
                    self._error_cb(e)

                time.sleep(self.retry_wait)
        # Don't listen to auth error, since we can't reasonably reconnect
        # when we get one.





tweetstream/streamclasses.py
import time
import json

import requests

from . import AuthenticationError, USER_AGENT
from . import ReconnectImmediatelyError, ReconnectLinearlyError, ReconnectExponentiallyError


class BaseStream(object):
    """A network connection to Twitters streaming API

    :param username: Twitter username for the account accessing the API.
    :param password: Twitter password for the account accessing the API.
    :keyword count: Number of tweets from the past to get before switching to
      live stream.
    :keyword raw: If True, return each tweet's raw data direct from the socket,
      without UTF8 decoding or parsing, rather than a parsed object. The
      default is False.
    :keyword timeout: If non-None, set a timeout in seconds on the receiving
      socket. Certain types of network problems (e.g., disconnecting a VPN)
      can cause the connection to hang, leading to indefinite blocking that
      requires kill -9 to resolve. Setting a timeout leads to an orderly
      shutdown in these cases. The default is None (i.e., no timeout).
    :keyword url: Endpoint URL for the object. Note: you should not
      need to edit this. It's present to make testing easier.

    .. attribute:: connected

        True if the object is currently connected to the stream.

    .. attribute:: url

        The URL to which the object is connected

    .. attribute:: starttime

        The timestamp, in seconds since the epoch, the object connected to the
        streaming api.

    .. attribute:: count

        The number of tweets that have been returned by the object.

    .. attribute:: rate

        The rate at which tweets have been returned from the object as a
        float. see also :attr: `rate_period`.

    .. attribute:: rate_period

        The ammount of time to sample tweets to calculate tweet rate. By
        default 10 seconds. Changes to this attribute will not be reflected
        until the next time the rate is calculated. The rate of tweets vary
        with time of day etc. so it's usefull to set this to something
        sensible.

    .. attribute:: user_agent

        User agent string that will be included in the request. NOTE: This can
        not be changed after the connection has been made. This property must
        thus be set before accessing the iterator. The default is set in
        :attr: `USER_AGENT`.
    """

    def __init__(self, username, password,
                 catchup=None, raw=False, timeout=None, url=None):
        self._conn = None
        self._rate_ts = None
        self._rate_cnt = 0
        self._username = username
        self._password = password
        self._catchup_count = catchup
        self._raw_mode = raw
        self._timeout = timeout
        self._iter = self.__iter__()

        self.rate_period = 10  # in seconds
        self.connected = False
        self.starttime = None
        self.count = 0
        self.rate = 0
        self.user_agent = USER_AGENT
        if url: self.url = url

        self._client = None

    def __enter__(self):
        return self

    def __exit__(self, *params):
        self.close()
        return False

    def _init_conn(self):
        """Open the connection to the twitter server"""

        if not self._client:
            headers = {'User-Agent': self.user_agent}
            auth = (self._username, self._password)
            config = {'danger_mode': True, 'keep_alive': False}
            self._client = requests.session(headers=headers,
                                            auth=auth,
                                            timeout=self._timeout,
                                            config=config)

        postdata = self._get_post_data() or {}
        if self._catchup_count:
            postdata["count"] = self._catchup_count

        req_method = 'post' if postdata else 'get'

        # If connecting fails, convert to ReconnectExponentiallyError so
        # clients can implement appropriate backoff.
        try:
            self._conn = self._client.request(req_method, self.url, data=postdata)
        except requests.HTTPError, e:
            code = e.response.status_code
            if code == 401:
                raise AuthenticationError("Access denied")
            elif code == 404:
                raise ReconnectExponentiallyError("%s: %s" % (self.url, e))
            else:
                raise ReconnectExponentiallyError(str(e))
        except requests.ConnectionError, e:
            raise ReconnectExponentiallyError(str(e))
        else:
            self.connected = True

        if not self.starttime:
            self.starttime = time.time()
        if not self._rate_ts:
            self._rate_ts = time.time()

    def _get_post_data(self):
        """Subclasses that need to add post data to the request can override
        this method and return post data. The data should be in the format
        returned by urllib.urlencode."""
        return None

    def _update_rate(self):
        rate_time = time.time() - self._rate_ts
        if not self._rate_ts or rate_time > self.rate_period:
            self.rate = self._rate_cnt / rate_time
            self._rate_cnt = 0
            self._rate_ts = time.time()

    def __iter__(self):
        # buf = b""
        while True:
            try:
                if not self.connected:
                    self._init_conn()

                for line in self._conn.iter_lines(chunk_size=1):
                    if (self._raw_mode):
                        tweet = line
                    else:
                        line = line.decode("utf8")
                        try:
                            tweet = json.loads(line)
                        except ValueError, e:
                            self.close()
                            raise ReconnectImmediatelyError("Got invalid data from twitter", details=line)

                    if 'text' in tweet:
                        self.count += 1
                        self._rate_cnt += 1
                    yield tweet

            except RuntimeError, e:
                self.close()
                raise ReconnectImmediatelyError("Server disconnected: %s" % (e, ))

    def next(self):
        """Return the next available tweet. This call is blocking!"""
        return self._iter.next()

    def close(self):
        """
        Close the connection to the streaming server.
        """
        self.connected = False
        if self._conn:
            self._conn.raw.release_conn()


class SampleStream(BaseStream):
    url = "https://stream.twitter.com/1/statuses/sample.json"


class FilterStream(BaseStream):
    url = "https://stream.twitter.com/1/statuses/filter.json"

    def __init__(self, username, password, follow=None, locations=None,
                 track=None, catchup=None, raw=False, timeout=None, url=None):
        self._follow = follow
        self._locations = locations
        self._track = track
        # remove follow, locations, track
        BaseStream.__init__(self, username, password,
                            raw=raw, timeout=timeout, url=url)

    def _get_post_data(self):
        postdata = {}
        if self._follow: postdata["follow"] = ",".join([str(e) for e in self._follow])
        if self._locations: postdata["locations"] = ",".join(self._locations)
        if self._track: postdata["track"] = ",".join(self._track)
        return postdata
